### PR TITLE
Port -tg-station cameranets

### PR DESCRIPTION
**Pull request by tigercat2000**
This should make it much easier to port abductors, if we ever want to do
that.

:cl:
Fixed a bug where the AI could not delete photos it takes.
Ported -tg- cameranets. This means that all of our camera based systems are more or less up to date with -tg-; 510 features are missing.
An AI with the 'camera lights' mode on will now, instead of the button toggling the closest camera's light, toggle camera lights on/off as the AI moves.
It now takes a minimum of 30 deciseconds between attempting to track someone, and locking on. This number is increased by their distance away from the AI eye.
The camera activate/deactivate proc is now called 'toggle_cam'.
The 'trackable' proc has been replaced by a more object-oriented 'mob.can_track()' proc.
The 'networks' section of the security camera console is now above the camera list.
/:cl:

In laymans terms: This refactors how security cameras calculate what
viewers can see, and adds a few neat features on top.

This does **not** port the -tg- advanced camera computers or the xenobiology computers; Those can come at a later date.

@Fox-McCloud

### CODE REVIEW

**Comment by Fox-McCloud** [comment]
Ahhh Yiiisss. 

[X] Organs
[ ] Cameranets
[ ] Diseases
[ ] Abductors

File: html/changelogs/Tigercat2000-tgcams.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+# Also, this gets changed to [] after reading.  Just remove the brackets when you add new shit.
+# Please surround your changes in  double quotes ("). It works without them, but if you use certain characters it screws up compiling. The quotes will not show up in the changelog.
+changes: 
+  - bugfix: "Fixed a bug where the AI could not delete photos it takes."
+  - rscadd: "Ported -tg- cameranets. This means that all of our camera based systems are more or less up to date with -tg-; 510 features are missing."
+  - rscadd: "An AI with the 'camera lights' mode on will now, instead of the button toggling the closest camera's light, toggle camera lights on/off as the AI moves."
+  - rscadd: "It now takes a minimum of 30 deciseconds between attempting to track someone, and locking on. This number is increased by their distance away from the AI eye."
+  - tweak: "The camera activate/deactivate proc is now called 'toggle_cam'."
+  - tweak: "The 'trackable' proc has been replaced by a more object-oriented 'mob.can_track()' proc."
+  - tweak: "The 'networks' section of the security camera console is now above the camera list."
```
Comment (Fox-McCloud) [inline html/changelogs/Tigercat2000-tgcams.yml]: .....no

Comment (tigercat2000) [inline html/changelogs/Tigercat2000-tgcams.yml]: .....yes. Rule 531 of UI design, things that change the height of the page should always be available at wherever the page resets the scroll to when the height changes.

File: code/game/machinery/camera/camera.dm [hunk 2]
```diff
@@ -76,26 +59,25 @@
 	wires = null
 	cameranet.removeCamera(src) //Will handle removal from the camera network and the chunks, so we don't need to worry about that
 	cameranet.cameras -= src
-	var/list/open_networks = difflist(network,restricted_camera_networks)
-	if(open_networks.len)
-		cameranet.removeCamera(src)
+	cameranet.removeCamera(src)
```
Comment (Fox-McCloud) [inline code/game/machinery/camera/camera.dm]: Uhhh, this seems like a mistake. I realize it's on TG, as well, but how can you remove a camera you just removed?

File: code/game/machinery/camera/tracking.dm [hunk 3]
```diff
@@ -108,53 +114,37 @@
...
-			continue
-		if(M.invisibility)//cloaked
-			continue
-		if(M.digitalcamo)
+		if(!M.can_track(usr))
 			continue
 
 		// Human check
 		var/human = 0
 		if(istype(M, /mob/living/carbon/human))
```
Comment (Fox-McCloud) [inline code/game/machinery/camera/tracking.dm]: my want to excluse `issmall` things.

File: code/modules/mob/living/silicon/ai/freelook/chunk.dm [hunk 4]
```diff
@@ -21,26 +21,24 @@
...
-		ai.ai.client.images -= obscured
-	seenby -= ai
+/datum/camerachunk/proc/remove(mob/camera/aiEye/eye)
+	var/client/client = eye.GetViewerClient()
+	if(client)
+		client.images -= obscured
+	eye.visibleCameraChunks -= src
+	seenby -= eye
 	if(visible > 0)
 		visible--
```
Comment (Fox-McCloud) [inline code/modules/mob/living/silicon/ai/freelook/chunk.dm]: this proc is a hair bit different on TG.

File: html/changelogs/Tigercat2000-tgcams.yml (hunk 1 above)
Comment (Fox-McCloud) [inline html/changelogs/Tigercat2000-tgcams.yml]: No, I meant this entire changelog, not this line specifically.

File: code/modules/mob/living/silicon/ai/freelook/eye.dm [hunk 5]
```diff
@@ -63,40 +33,30 @@
...
 /atom/proc/move_camera_by_click()
 	if(istype(usr, /mob/living/silicon/ai))
 		var/mob/living/silicon/ai/AI = usr
 		if(AI.eyeobj && AI.client.eye == AI.eyeobj)
-			AI.eyeobj.setLoc(src)
+			AI.cameraFollow = null
+			if (isturf(src.loc) || isturf(src))
+				AI.eyeobj.setLoc(src)
+
+/mob/aiEye/experience_pressure_difference()
```
Comment (Fox-McCloud) [inline code/modules/mob/living/silicon/ai/freelook/eye.dm]: Does this type even exist anymore?

File: html/changelogs/Tigercat2000-tgcams.yml (hunk 1 above)
Comment (tigercat2000) [inline html/changelogs/Tigercat2000-tgcams.yml]: @Fox-McCloud put it in the contributing file if it's not allowed, otherwise, using this

File: code/game/machinery/camera/camera.dm (hunk 2 above)
Comment (tigercat2000) [inline code/game/machinery/camera/camera.dm]: I have no idea, yeah.

File: code/modules/mob/living/silicon/ai/freelook/eye.dm (hunk 5 above)
Comment (Spacemanspark) [inline code/modules/mob/living/silicon/ai/freelook/eye.dm]: You can follow it in the follow verb while a ghost, so I assume so.

**Comment by TheDZD** [comment]
untitled
@tigercat2000

File: code/game/machinery/camera/tracking.dm (hunk 3 above)
Comment (tigercat2000) [inline code/game/machinery/camera/tracking.dm]: eh, it was always like this, would rather not fuck with it.

File: code/modules/mob/living/silicon/ai/freelook/chunk.dm (hunk 4 above)
Comment (tigercat2000) [inline code/modules/mob/living/silicon/ai/freelook/chunk.dm]: https://github.com/tgstation/-tg-station/blob/master/code/modules/mob/living/silicon/ai/freelook/chunk.dm#L36-L43
```
/datum/camerachunk/proc/remove(mob/camera/aiEye/eye)
	var/client/client = eye.GetViewerClient()
	if(client)
		client.images -= obscured
	eye.visibleCameraChunks -= src
	seenby -= eye
	if(visible > 0)
		visible--
```
https://github.com/ParadiseSS13/Paradise/pull/3865/files#diff-fb6a7279dcd9db54d57ca7d0aa88fbcaR36
```
/datum/camerachunk/proc/remove(mob/camera/aiEye/eye)
	var/client/client = eye.GetViewerClient()
	if(client)
		client.images -= obscured
	eye.visibleCameraChunks -= src
	seenby -= eye
 	if(visible > 0)
 		visible--
```

**Comment by tigercat2000** [comment]
who the fuck conflicted this

File: code/modules/mob/living/silicon/ai/freelook/chunk.dm (hunk 4 above)
Comment (Fox-McCloud) [inline code/modules/mob/living/silicon/ai/freelook/chunk.dm]: wha-...ok. Weird.

File: code/game/machinery/camera/tracking.dm (hunk 3 above)
Comment (Fox-McCloud) [inline code/game/machinery/camera/tracking.dm]: well, it was more uh, monkies being carbons on TG and being excluded. It's not a big deal, was just thinking about it, s'all.

**Comment by tigercat2000** [comment]
you fucking wot mate

**Comment by Fox-McCloud** [comment]
looks like the mob helper for the AI Eye needs to be changed.

Also, yeah, AI detection multitool will probably need to be updated--forgot to mention that.
